### PR TITLE
Add tcpudp plugin

### PR DESCRIPTION
**Pull request by ericwholt**
Add plugin to use tcp_input without any extra parsing.

### CODE REVIEW

File: plugins/tcpudp.yaml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+version: 0.0.1
+title: TCP/UDP
+description: TCP/UDP Input Parser
+min_stanza_version: 1.2.7
+parameters:
+  - name: listen_port
+    label: Listen Port
+    description: A port which the agent will listen for syslog messages
+    type: int
+    default: 5140
```
Comment (jsirianni) [inline plugins/tcpudp.yaml:11]: I think we should remove default value for port and make it a required value.

File: plugins/tcpudp.yaml [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    type: enum
+    valid_values:
+      - udp
+      - tcp
+    default: udp
+  - name: log_type
+    label: Type
+    description: Adds the specified 'Type' as a label to each log message.
+    type: string
+    default: tcpudp
```
Comment (jsirianni) [inline plugins/tcpudp.yaml:24]: I think this should be conditional based on the connection_type

File: plugins/tcpudp.yaml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  - name: listen_ip
+    label: Listen IP
+    description: A syslog ip address of the form `<ip>`
```
Comment (jsirianni) [inline plugins/tcpudp.yaml:27]: `syslog`

File: plugins/tcpudp.yaml [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    description: Adds net.transport, net.peer.ip, net.peer.port, net.host.ip and net.host.port labels.
+    type: bool
+    default: true
+    advanced_config: true
+  - name: enable_tls
+    label: Enable TLS
+    description: Enable TLS for the TCP listener
+    type: bool
+    default: false
+    advanced_config: true
```
Comment (jsirianni) [inline plugins/tcpudp.yaml:42]: Only relevant for TCP

File: plugins/tcpudp.yaml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    description: File path for the X509 TLS certificate chain
+    type: string
+    default: "/opt/key"
+    relevant_if:
+      enable_tls:
+        equals: true
+    advanced_config: true
+# Set Defaults
+# {{$listen_ip := default "0.0.0.0" .listen_ip}}
+# {{$listen_port := default 514 .listen_port}}
```
Comment (jsirianni) [inline plugins/tcpudp.yaml:63]: default port does not match input param config

File: plugins/tcpudp.yaml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      plugin_id: {{ .id }}
+    add_labels: {{$add_labels}}
+    # {{ if $enable_tls}}
+    tls:
+      enable: {{ $enable_tls }}
+      certificate: {{ $certificate_file }}
+      private_key: {{ $private_key_file }}
+    # {{ end }}
+    output: {{.output}}
+# {{ end }}
```
Comment (jsirianni) [inline plugins/tcpudp.yaml:87]: UDP cannot use TLS, as TLS inherently requires a TCP connection. There are situations where one could pass UDP over a TLS connection but Stanza does not support it.

File: plugins/tcpudp.yaml [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+# {{$certificate_file := default "" .certificate_file}}
+# {{$private_key_file := default "" .private_key_file}}
+# Pipeline Template
+pipeline:
+# {{ if eq $connection_type "udp" }}
+  - id: upd_input
+    type: udp_input
+    listen_address: '{{ $listen_ip }}:{{ $listen_port }}'
+    labels:
+      log_type: syslog
```
Comment (jsirianni) [inline plugins/tcpudp.yaml:77]: wrong log_type

File: plugins/tcpudp.yaml [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    # {{ end }}
+    output: {{.output}}
+# {{ end }}
+
+# {{ if eq $connection_type "tcp" }}
+  - id: tcp_input
+    type: tcp_input
+    listen_address: '{{ $listen_ip }}:{{ $listen_port }}'
+    labels:
+      log_type: syslog
```
Comment (jsirianni) [inline plugins/tcpudp.yaml:94]: wrong log_type

**Review by jsirianni — CHANGES_REQUESTED** [review]

**Comment by jsirianni** [comment]
Lets go ahead and split this into two plugins

File: plugins/tcpudp.yaml (hunk 2 above)
Comment (ericwholt) [reply plugins/tcpudp.yaml:24]: We are splitting this plugin into two separate plugins tcp and udp. I will set default to either tcp or udp respectively.

File: plugins/tcpudp.yaml (hunk 5 above)
Comment (ericwholt) [reply plugins/tcpudp.yaml:63]: Removed as we now require port

File: plugins/tcpudp.yaml (hunk 6 above)
Comment (ericwholt) [reply plugins/tcpudp.yaml:87]: done

File: plugins/tcpudp.yaml (hunk 7 above)
Comment (ericwholt) [reply plugins/tcpudp.yaml:77]: done

File: plugins/tcpudp.yaml (hunk 8 above)
Comment (ericwholt) [reply plugins/tcpudp.yaml:94]: done

File: plugins/tcp.yaml [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  - name: listen_port
+    label: Listen Port
+    description: A port which the agent will listen for tcp messages
+    type: int
+    required: true
+  - name: log_type
+    label: Type
+    description: Adds the specified 'Type' as a label to each log message.
+    type: string
+    default: tcpudp
```
Comment (jsirianni) [inline plugins/tcp.yaml:16]:
```suggestion
    default: tcp
```

File: plugins/tcp.yaml [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    label: Listen IP
+    description: A ip address of the form `<ip>`
+    type: string
+    default: "0.0.0.0"
+    advanced_config: true
+  - name: add_labels
+    label: Labels
+    description: Adds net.transport, net.peer.ip, net.peer.port, net.host.ip and net.host.port labels.
+    type: bool
+    default: tcp
```
Comment (jsirianni) [inline plugins/tcp.yaml:27]:
```suggestion
    default: true
```

File: plugins/udp.yaml [hunk 11]
```diff
@@ -0,0 +1,44 @@
+# Plugin Info
+version: 0.0.1
+title: TCP/UDP
+description: TCP/UDP Input Parser
```
Comment (jsirianni) [inline plugins/udp.yaml:4]:
```suggestion
description: UDP Input
```

File: plugins/udp.yaml [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    label: Labels
+    description: Adds net.transport, net.peer.ip, net.peer.port, net.host.ip and net.host.port labels.
+    type: bool
+    default: true
+    advanced_config: true
+
+# Set Defaults
+# {{$listen_port := .listen_port}}
+# {{$listen_ip := default "0.0.0.0" .listen_ip}}
+# {{$log_type := default "tcpudp" .log_type}}
```
Comment (jsirianni) [inline plugins/udp.yaml:33]:
```suggestion
# {{$log_type := default "udp" .log_type}}
```

File: test/configs/tcp/invalid/invalid_listen_port.yaml [hunk 13]
```diff
@@ -0,0 +1,5 @@
+pipeline:
+- type: tcp
+  listen_port: 6543
```
Comment (jsirianni) [inline test/configs/tcp/invalid/invalid_listen_port.yaml:3]:
```suggestion
```

File: plugins/udp.yaml [hunk 14]
```diff
@@ -0,0 +1,44 @@
+# Plugin Info
+version: 0.0.1
+title: TCP/UDP
```
Comment (jsirianni) [inline plugins/udp.yaml:3]:
```suggestion
title: UDP
```

**Review by jsirianni — APPROVED** [review]
```json
{
  "timestamp": "2021-09-29T12:39:27.55964-04:00",
  "severity": 0,
  "labels": {
    "log_type": "udp",
    "net.host.ip": "::",
    "net.host.port": "5001",
    "net.peer.ip": "::1",
    "net.peer.port": "63507",
    "net.transport": "IP.UDP",
    "plugin_id": "udp"
  },
  "record": "hi"
}
{
  "timestamp": "2021-09-29T12:39:32.870037-04:00",
  "severity": 0,
  "labels": {
    "log_type": "tcp",
    "net.host.ip": "::1",
    "net.host.port": "5000",
    "net.peer.ip": "::1",
    "net.peer.port": "57838",
    "net.transport": "IP.TCP",
    "plugin_id": "tcp"
  },
  "record": "hi"
}
```

**Pull request merged**